### DISCUSSION TITLE
Force direction detected by sensor

### DISCUSSION THREAD

**Discussion by zichunxx** [google-deepmind/mujoco#2006] [Asking for Help]
Hi! 

I've been confused about the force direction detected by the force sensor for a while and made a [MWE.zip](https://github.com/google-deepmind/mujoco/files/14392222/MWE.zip) to illustrated my confusion with the [impedance controller](https://github.com/kevinzakka/mjctrl) made by @kevinzakka.

First of all, I found that creating a dummy body is a great way to gather force information from [here](https://github.com/google-deepmind/mujoco/issues/399#issue-1317892443).

Second, as illustrated in the [document](https://mujoco.readthedocs.io/en/stable/XMLreference.html#sensor-force), the detected force is expressed in the site frame and points from the child body to the parent body.

For the above MWE, I create a dummy body containing a sensor site on the tip of the peg and simulate a process of moving vertically downward toward the ground and hitting the ground. **Running the python script in the MWE can reproduce the process directly without any tuning.**

<img src="https://github.com/google-deepmind/mujoco/assets/72594017/c33622f6-bb0f-4448-ba59-b2e99569c8c4" width=50% height=50%>

Here is the force information in the z-direction:
<img src="https://github.com/google-deepmind/mujoco/assets/72594017/1352c8b2-354b-4fd4-af36-a3f397f32985" width=60% height=60%>

For my example, I think the force is negative because the z-axis of the sensor site frame points downward, and the force points from the dummy body to the box_eef body, which is upward. You can also find body labels in the above figure.

1) But why the detected force is positive? 

2) I also found the dummy body should be in the right position but the position of the sensor site does not matter, why?

I have found some related issues but they didn't solve my confusion. Here is the list:

* https://github.com/google-deepmind/mujoco/issues/1249
* https://github.com/google-deepmind/mujoco/issues/399
* https://github.com/google-deepmind/mujoco/issues/997

Thanks in advance for your precious time in checking the MWE. Any comment would be appreciated.

**Comment by kevinzakka** [comment]
Hi @zichunxx, I'm looking into this and will get back to you soon.

**Comment by zichunxx** [comment]
> Hi @zichunxx, I'm looking into this and will get back to you soon.

Thanks in advance for your generous help.

**Comment by zichunxx** [comment]
Hi! @kevinzakka Sorry for the interruption. Did you find anything wrong?

**Comment by egordon** [comment]
Chiming in a little late, but I am experiencing something similar.

Loading this XML:
[default.txt](https://github.com/user-attachments/files/21448881/default.txt)

`spherebot1` is on the `+x` side of the box, while `spherebot2` is on the `-x` side of the box. They should start in collision.

When I check the contact sensor output using the following code (`v3.3.4`) it appears to return the opposite of what I would expect given the output of `mj_contactForce`. I apologize if I have some misunderstanding, but at the moment in more complicated code I just have to take `-1 * sensordata`.

```
import mujoco
import numpy as np
mj_model = mujoco.MjModel.from_xml_path("default.txt")
mj_data = mujoco.MjData(mj_model)
mujoco.mj_step(mj_model, mj_data)
print(f"Spherebot1 [+x] -> Box Geometry ID: {mj_data.geom('spherebot1-geom').id} -> {mj_data.geom('object-geom').id}") # 2 -> 1
print(f"Contact 4 Geometry ID: {mj_data.contact[4].geom}") # [2 1]
result = np.zeros(6)
mujoco.mj_contactForce(mj_model, mj_data, 4, result)
world_frame_force_on_box = mj_data.contact[4].frame.reshape((3,3)).T @ result[:3].T
# Note: diagonal matrix in this case so removing '.T' doesn't change result

print(f"Contact force on Box (world frame): {world_frame_force_on_box}") # [-4... 0... 2....]
print(f"Contact force on Spherebot1 (world frame): {-world_frame_force_on_box}") # [4... 0... -2....]
print(f"Site Frame (== world frame): {mj_data.site("spherebot1-site").xmat.reshape((3,3))}") # Identity

#### This seems backwards
start = mj_model.sensor("spherebot1-contact").adr[0]
end = start + mj_model.sensor("spherebot1-contact").dim[0]
print(f"Spherebot1 contact sensordata (site == world frame): {mj_data.sensordata[start:end]}") # [-4... 0... 2....]
####
```

**Comment by prasuchit** [comment]
I realize it's been a while since this was posted but I've run into the same confusion now. I have a force-torque sensor attached to the end-effector of my robot and when I push against an object (say from left +Y to right -Y on the object's left wall), the real robot's sensor gives me the reactive force (similar to a strain-gauge, as I would expect) along +Y (transformed to world-frame). However, when I do the same thing with my sim robot, I see the force in -Y (also transformed to world-frame).

The confusing part is that[ Mujoco docs say](https://mujoco.readthedocs.io/en/stable/XMLreference.html#sensor-force) its sensors measure:
`The sensor outputs three numbers, which are the interaction force between a child and a parent body, expressed in the site frame defining the sensor. The convention is that the site is attached to the child body, and the force points from the child towards the parent. `

I'm currently using MJLab, so perhaps @kevinzakka could help me with this? Thanks in advance.
